### PR TITLE
Refactor/computed observers

### PR DESCRIPTION
**Pull request by bigopon**
# Pull Request

## 📖 Description

Rework getter observation, based on new proxy implementation & add some more (unit) tests close to the lower level of the watcher switcher, and the proxy observable APIs

## 📑 Test Plan

Focus on the APIs of the switcher & proxy observable, but still keep watch decorator as the main driver for proxy tests

## ⏭ Next Steps

More observation refactoring related to bindable observer & typing of the observation

### CODE REVIEW

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://59602-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by bigopon** [comment]
For `IWatchDefinition`, I think we should skip the part of making it a class. I'm not sure the benefit outweigh the simplicity loss, and the slightly bigger bundle size. Also it shares the same style with our bindables (definition places, and the  configuration construction)

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://59792-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://59853-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://59859-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://59963-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://59973-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://60056-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

File: packages/runtime/src/observation/watcher-switcher.ts [hunk 1]
```diff
@@ -64,3 +64,16 @@ export function exitWatcher(watcher: IWatcher): void {
   $watcher = watchers.length > 0 ? watchers[watchers.length - 1] : null;
   watching = $watcher != null;
 }
+
+export const WatcherSwitcher = Object.freeze({
```
Comment (fkleuver) [inline packages/runtime/src/observation/watcher-switcher.ts:68]: Why the double export? We could also expose `$watcher` via a `getCurrentWatcher()` fn etc right?

File: packages/runtime/src/observation/proxy-observation.ts [hunk 2]
```diff
@@ -1,13 +1,30 @@
 import { IIndexable } from '@aurelia/kernel';
-import { watching, currentWatcher } from './watcher-switcher';
+import { watching, currentWatcher } from './watcher-switcher.js';
 
 const R$get = Reflect.get;
+const toStringTag = Object.prototype.toString;
 const proxyMap = new WeakMap<object, object>();
 
+function toRawType(value: unknown): string {
+  return toStringTag.call(value).slice(8, -1);
```
Comment (fkleuver) [inline packages/runtime/src/observation/proxy-observation.ts:9]: This isn't more efficient than just comparing to `[object xxxx]`

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://60098-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

File: packages/testing/src/test-builder.ts [hunk 3]
```diff
@@ -464,6 +464,7 @@ export function createObserverLocator(containerOrLifecycle?: IContainer | ILifec
       return false;
     }
   };
+  Registration.instance(IPlatform, {}).register(container);
```
Comment (fkleuver) [inline packages/testing/src/test-builder.ts:467]: This shouldn't be needed. If it is, we need to initialize a TestContext earlier.

I want to get rid of these types of no-op registrations, not add more of them. They make things flaky.

File: packages/runtime/src/observation/watcher-switcher.ts (hunk 1 above)
Comment (bigopon) [reply packages/runtime/src/observation/watcher-switcher.ts:68]: This is what I'm not sure about. Should everything related to watch be grouped, and the loose exports will be used internally inside runtime only, or should the external packages use the same?

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://60115-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://60152-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://60275-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://60317-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Pull request merged**